Add conan v1 cheatsheet (#2071)
conan-io/docs: cheatsheet.rst

Install configurations:

.. code-block:: bash

    $ conan config install <item>  # Copy the relevant contents from <item> to the user's ~/.conan directory.

    $ conan config install ./my_config.conf

Alternatively, copying files and editing conan.conf can be done manually.

Set up configurations:

.. code-block:: bash

    $ conan config init     # Initialize Conan configuration files. If some are already present, missing files only are
                            # created

Set configuration values:

.. code-block:: bash

    $ conan config set <section>.<config>=<value>

    $ conan config set log.level=10
    $ conan config set log.print_run_commands=False  # Make conan less verbose

Inspect configurations:

.. code-block:: bash

    $ conan config home     # See the Conan home directory

    $ conan config get [<section>.<config>]  # Show some or all configuration items

    $ conan config get                       # Show the full conan.conf file
    $ conan config get log.level             # Show the "level" item in the "log" section

See `conan config <https://docs.conan.io/en/latest/reference/commands/consumer/config.html>`_ reference.

Profiles
++++++++

> Profiles allow users to set aspects of the build environment. This includes settings_, options_, environment variables

Author: Sorry could you clarify - my understanding was that profiles only cover settings, options, environment variables and build requirements: https://docs.conan.io/en/latest/reference/profiles.html

> Profiles allow users to set aspects of the build environment. This includes settings_, options_, environment variables

Author: The "conf_" has currently been left out, let me know if this needs changing

        generators = "cmake"                               # Generator for the package: specifies which build system
                                                           # type will be generated

        def source(self):                                                # Obtains the source code for the project
            self.run("git clone https://github.com/conan-io/hello.git")  # self.run() executes any command in the native shell
            ...

        def build(self):                                                 # Responsible for invoking the build system
            cmake = CMake(self)                                          # Helper classes are available for several build systems
            ...
            if self.options.myoption1:                                   # Specify a conditional build requirement
                self.build_requires("zlib/1.2@user/testing")
            self.run("bin/unittests")                                    # Run unit tests compiled earlier in the build() method

        def package(self):                                               # Responsible for capturing build artifacts

>             self.copy("\*.h", dst="include", src="hello")                # self.copy() copies files from the cache to the project folder

Author: IIRC the backslash is there to escape the asterisk

>             self.copy("\*.h", dst="include", src="hello")                # self.copy() copies files from the cache to the project folder

Reviewer: it is not needed.

            ...

        def package_info(self):                                          # Responsible for defining variables that are
                                                                         # passed to package consumers, for example
                                                                         # library or include directories
            self.cpp_info.libs = ["hello"]                               # The cpp_info dictionary contains these variables
            ...

        def requirements(self):                                          # Responsible for specifying non-trivial requirements logic
            if self.options.myoption2:                                   # Specify a conditional requirement

>                 self.requires("RequiredLib2/0.3@user/stable")

Author: I prefer to keep both, if you don't object, because it tells the user that the methods exists

>         def package_id(self):                                            # Responsible for changing the way the package

Reviewer: ```suggestion
        def package_id(self):                                            # Responsible for overriding the way the package
```

>                                                                          # ID is calculated from the default

Author: "while building this package only"? Because as far as I understand, you can affect the package ID modes of dependencies using this method, e.g. `self.info.requires[dep].recipe_revision_mode()`. Correct me if I'm wrong...
(Aside: what happens if `dep` defines a different package ID mode for itself, in its recipe? Does the mode defined in the dependent package win?)

>                                                                          # ID is calculated from the default

Reviewer: highly advanced question, don't know off hand, maybe not a good candidate for this "cheatsheet"

>                                                                          # ID is calculated from the default

Author: Ok thanks. This sentence now reads "Responsible for overriding the way the package ID is calculated from the default, for this package only"

__ #creating-a-basic-package

Options
#######

Options are configuration items which are generally package-specific.

The available options for a package are defined in `the package recipe`_.

Options can then be set via profiles_, an application's `Conanfile.txt`__, or via arguments to `conan install`__ or
`conan create`__.

__ #using-conan-packages-in-an-application
__ #using-conan-packages-in-an-application
__ #creating-a-basic-package

Versioning
++++++++++

Versions
########

Packages are specified whenever a package is created, and whenever a recipe is consumed via a recipe reference.

Specify ranges:

.. code-block:: text

    [>min_ver <max_ver] - specify a version range
    [*]                 - specify any version
    [~maj.min]          - specify any patch in v[maj].[min]

The version taken is otherwise the maximum available.

Revisions
#########

Revisions allow changes to a package without increasing the version number or overwriting the existing version number.
They are disabled by default.

There are two types of revisions:

- "Recipe Revisions" (RREV) - Revision of the recipe and sources
- "Package Revisions" (PREV) - Revision of a binary package

> The recipe revision (RREV) is a SHA-1 hash either calculated over the recipe, or taken from the version control system.

Reviewer: ```suggestion
The recipe revision (RREV) is a SHA-1 hash calculated over the `conan_manifest.txt`, which contains the individual hashes of the `conanfile.py`, plus all the files exported with `exports` and `exports_sources`. 
If the `scm` feature is used, Conan can also formulate the recipe revision directly from the version control system.
```

Conan only holds one recipe revision in the local cache. Many recipe revisions can be stored in remote repositories.
This helps differentiate between packages that have been changed and built without changing the version number. Recipe
revisions can be specified wherever a recipe is consumed. If a recipe revision is not specified, the latest revision is
used.

> The package revision (PREV) is a SHA-1 hash calculated over the package contents. Package revisions provide the most

Reviewer: ```suggestion
The package revision (PREV) is a SHA-1 hash calculated over the binary contents of the package 
directory after the build and package steps are completed. Package revisions provide the most
```

Specifying dependencies
#######################

Main application dependencies are set in the [requires] section of `Conanfile.txt`__.

__ #using-conan-packages-in-an-application

Package dependencies - normal requirements, build requirements, conditional requirements - are set in `the package recipe`_.

Package ID calculation modes
############################

Conan performs dependency resolution via the calculation of package IDs. A package ID is calculated for a desired
dependency, and then Conan searches for that package ID.

The package ID calculation, and therefore the dependency resolution, is affected by the default_package_id_mode and the
default_python_requires_id_mode. They determine what exactly affects the calculation: which parts of version numbers;
package revisions; immediate or transitive dependencies. This relates to both normal requirements and `Python
requires`_. By default, only the main version number of direct dependencies are taken into account when calculating the
package ID.

These modes can be set in the [general] section of configurations_, and in `the package recipe`_.

Resolving dependency conflicts
##############################

> Versions defined in the `Conanfile.txt`__ take precedence over versions specified by dependencies. This can be used to

Reviewer: ```suggestion
Versions defined in the `conanfile.txt`__ take precedence over versions specified by dependencies. This can be used to
```
